Issue 4136

atavism opened the pull request:
This resolves https://github.com/getlantern/lantern/issues/4136 

With Lantern running, if you change your device locale and bring the app back to the foreground, the feed in the new locale will automatically load.

This updates the feed ViewPager to use a FragmentStatePagerItemAdapter instead of a FragmentPagerItemAdapter. The former allows us to preserve state and dynamically update feed headers in the tab layout.

myleshorton commented on src/github.com/getlantern/lantern-mobile/app/src/main/java/org/getlantern/lantern/fragment/FeedFragment.java:
```diff
@@ -31,29 +31,22 @@
     @Override
     public void onCreate(Bundle savedInstanceState) {
         super.onCreate(savedInstanceState);
+        this.mFeedItems = new ArrayList<FeedItem>();
```
Doesn't this still need to be a synchronized list @atavism?

atavism replied:
No. We only add elements to the feed items list now in ``onPostExecute`` of the AsyncTask (this is invoked on the UI thread).

myleshorton commented on src/github.com/getlantern/lantern-mobile/app/src/main/java/org/getlantern/lantern/fragment/FeedFragment.java:
```diff
@@ -62,28 +55,48 @@ public View onCreateView(LayoutInflater inflater, ViewGroup container,
+        @Override
+        protected void onPostExecute(List<FeedItem> items) {
+            super.onPostExecute(items);
+            mFeedItems.clear();
```
I believe you need to separately synchronize clear and addAll here @atavism.

myleshorton replied:
Or potentially use a copyOnWriteArrayList

atavism replied:
It doesn't need to be synchronized any longer (that's why I moved the processing inside an AsyncTask). It happens on the UI thread now.

myleshorton replied:
Ah cool I see that now - nice @atavism.

myleshorton commented on src/github.com/getlantern/lantern-mobile/app/src/main/java/org/getlantern/lantern/fragment/FeedFragment.java:
```diff
@@ -31,29 +31,22 @@
-        });
-    }   
+    public void setFeedName(String feedName) {
+        this.feedName = feedName;
```
This is called automatically by the JSON parser or something @atavism is that right?

myleshorton commented on src/github.com/getlantern/lantern-mobile/app/src/main/java/org/getlantern/lantern/fragment/FeedFragment.java:
```diff
@@ -62,28 +55,48 @@ public View onCreateView(LayoutInflater inflater, ViewGroup container,
-            }
-        }.start();
+        Log.d(TAG, "onViewCreated for " + this.feedName);
+        new LoadFeed().execute(this.feedName);
```
I'm worried about the feedName being potentially null or "" here @atavism. Is that possible?

Also, a unit test would be pretty amazing here.

atavism replied:
Yeah, definitely agree we should add a unit test. I'm actually working on improving test coverage overall on a separate branch so I made a note of it.

myleshorton commented on src/github.com/getlantern/lantern-mobile/app/src/main/java/org/getlantern/lantern/fragment/FeedFragment.java:
```diff
@@ -62,28 +55,48 @@ public View onCreateView(LayoutInflater inflater, ViewGroup container,
         return view;
     }
 
+    protected class LoadFeed extends AsyncTask<String, Void, List<FeedItem>> {
```
Seems like a private class would make more sense here @atavism -- i.e. I don't think we're ever going to subclass this and nothing else in this package will user, or maybe more precisely I don't think we want to allow either of those to happen.

atavism replied:
Agreed. No reason for the class not to be made it private. Should be updated now!

myleshorton commented on src/github.com/getlantern/lantern-mobile/app/src/main/java/org/getlantern/lantern/fragment/FeedFragment.java:
```diff
@@ -31,29 +31,22 @@
-        this.feedName = bundle.getString("name");
-        this.mFeedItems = Collections.synchronizedList(new ArrayList<FeedItem>());
+        if (bundle != null) {
+            this.feedName = bundle.getString("name");
```
Yeah I'm worried here about the feed name never getting set. Ideally to me this would be an invariant of the class -- i.e. make feedName `final` and check any arguments assigning it for validity.

myleshorton replied:
And ditch `setFeedName` if possible -- make it immutable.

atavism replied:
Whenever the feed fragment is created, we're always passing in a bundle with a name argument. There's also multiple checks before that to make sure we have a valid name. That said, I'm returning an empty list of feed items now whenever we encounter a null or empty string name. I also removed setFeedName (wasn't being used anyway).

myleshorton commented:
Hey @atavism I feel like this one actually isn't critical to 2.2.0, but please correct me if I'm wrong. I'm going ahead and building a new 2.2.0 now, but maybe we can just shift this to 2.2.1 or something.

atavism commented on src/github.com/getlantern/lantern-mobile/app/src/main/java/org/getlantern/lantern/fragment/FeedFragment.java:
```diff
@@ -24,66 +24,79 @@
 
     @Override
     public void onCreate(Bundle savedInstanceState) {
         super.onCreate(savedInstanceState);
```
Whenever the feed fragment is created, we're always passing in a bundle with a name argument. There's also multiple checks before that to make sure we have a valid name. That said, I'm returning an empty list of feed items now whenever we encounter a null or empty string name. I also removed ``setFeedName`` (wasn't being used anyway).

atavism commented:
@myleshorton Not super critical, but without this, we can't change the device locale to load different feeds without restarting the app. I addressed your comments and would say we should try to get it in if we can.

myleshorton commented:
OK LGTM @atavism!

atavism commented:
Thanks, @myleshorton!
